src/backend/distributed/utils/metadata_cache.c

```diff
@@ -2912,6 +2934,51 @@ InvalidateDistRelationCacheCallback(Datum argument, Oid relationId)
 }
 
 
+/*
+ * SetRelatedOidsInvalid sets the affected relation's cache entry to invalid.
+ * It also invalidates foreign key relation graph if any relation
+ */
+static void
+SetRelatedOidsInvalid(Oid relationId)
+{
+	List *connectedComponentListOfRelation = NIL;
+	ListCell *nodeCell = NULL;
+	bool foundInCache = false;
+
+	/*
+	 * We need to invalidate cache entry of each relation which are in the
+	 * same connected component with the given relation. If we do not have
+	 * foreign key relation graph, we don't need to do anything.
+	 *
+	 * We also need to invalidate foreign key relation graph if any of the
+	 * node in it is affected by invalidating given relation ID.
+	 */
+	if (IsForeignKeyGraphValid())
+	{
+		connectedComponentListOfRelation = ConnectedComponentOfRelationId(
+			relationId);
+
+		foreach(nodeCell, connectedComponentListOfRelation)
+		{
+			Oid neighbourRelationId = lfirst_oid(nodeCell);
+			void *neighbourHashKey = (void *) &neighbourRelationId;
+			DistTableCacheEntry *neighbourCacheEntry = hash_search(
+				DistTableCacheHash, neighbourHashKey, HASH_FIND, &foundInCache);
+
+			if (foundInCache)
+			{
+				neighbourCacheEntry->isValid = false;
+			}
+		}
+
+		/*if (list_length(connectedComponentListOfRelation) > 0) */
+		/*{ */
+		SetForeignKeyGraphInvalid();
```

Comment:
I think it'd be more clear to make this as a separate call after `SetRelatedOidsInvalid`.